### PR TITLE
Add ability to customize SearchResult children via SidebarSearchModal component

### PR DESCRIPTION
**Pull request by joshjung**
## Hey, I just made a Pull Request!

Adding ability to customize the children of the SearchResult component via the SidebarSearchModal component.

Screenshot 2024-08-16 at 11 18 56

#### :heavy_check_mark: Checklist

- [x] A changeset describing the change and affected packages. ([more info](https://github.com/backstage/backstage/blob/master/CONTRIBUTING.md#creating-changesets))
- [x] Added or updated documentation
- [x] Tests for new functionality and regression tests for bug fixes
- [x] Screenshots attached (for UI changes)
- [x] All your commits have a `Signed-off-by` line in the message. ([more info](https://github.com/backstage/backstage/blob/master/CONTRIBUTING.md#developer-certificate-of-origin))

### CODE REVIEW

**Comment by jvanalstyne-mdsol** [comment]
Thank you for working on this! I have been trying to implement a custom `SearchResultLinkItem` in the `SearchModal`, and went as far as trying to make a custom `SearchModal` as well, but it was way too complicated due to all the React hooks being used.

Your solution is much simpler - would love to see this accepted!

**Comment by joshjung** [comment]
@benjdlambert @freben Not sure why this PR not get someone assigned to review it!

**Comment by benjdlambert** [comment]
@joshjung oh yep this got missed for some reason. Could have been that GitHub was having a wobble at the time as it didn’t add anyone of the code owners at all. Added the correct groups manually, will take a look!

File: docs/features/search/how-to-guides.md
Comment (camilaibs) [inline docs/features/search/how-to-guides.md:301]: Prefixing the property name with search seems redundant. I suggest renaming it to `resultItems`.

File: plugins/search/src/components/SearchModal/SearchModal.tsx
Comment (camilaibs) [inline plugins/search/src/components/SearchModal/SearchModal.tsx:45]: It's not necessary to forward the search results here, because people who define a children function are following the third use case, they are completely customizing the modal.
```suggestion
export interface SearchModalChildrenProps {
```

File: plugins/search/src/components/SearchModal/SearchModal.tsx
Comment (camilaibs) [inline plugins/search/src/components/SearchModal/SearchModal.tsx:55]: Here is the only place where I would recommend defining the `resultItems` property.
```suggestion
export interface SearchModalProps extends SearchResultChildrenProvider {
  /**
   * Optional search result item extensions list.
   */
  resultItems?: JSX.Element | ((resultSet: SearchResultSet) => JSX.Element);
```

File: plugins/search/src/components/SearchModal/SearchModal.tsx
Comment (camilaibs) [inline plugins/search/src/components/SearchModal/SearchModal.tsx:105]:
```suggestion
  resultItems,
```

File: plugins/search/src/components/SearchModal/SearchModal.tsx
Comment (camilaibs) [inline plugins/search/src/components/SearchModal/SearchModal.tsx:170]:
```suggestion
          {resultItems}
```

File: plugins/search/src/components/SearchModal/SearchModal.tsx
Comment (camilaibs) [inline plugins/search/src/components/SearchModal/SearchModal.tsx:216]: As mentioned early in the comments, this is the case where people are completely customizing the search modal, they will define the search results items manually, so we don't need to forward result items.
```suggestion

```

File: plugins/search/src/components/SearchModal/SearchModal.tsx
Comment (camilaibs) [inline plugins/search/src/components/SearchModal/SearchModal.tsx:220]: The `SearchResult` component will use the default result item component if the `resultItems` is undefined:
```suggestion
              resultItems={resultItems}
```

File: plugins/search/src/components/SearchType/SearchType.Tabs.tsx
Comment (camilaibs) [inline plugins/search/src/components/SearchType/SearchType.Tabs.tsx:23]: Could you give context why this change was necessary?

File: plugins/search/src/components/SidebarSearchModal/SidebarSearchModal.tsx
Comment (camilaibs) [inline plugins/search/src/components/SidebarSearchModal/SidebarSearchModal.tsx:26]:
```suggestion

```

File: plugins/search/src/components/SidebarSearchModal/SidebarSearchModal.tsx
Comment (camilaibs) [inline plugins/search/src/components/SidebarSearchModal/SidebarSearchModal.tsx:36]: We can pick the `resultItem` type from the `SearchModalProps` 🙂 :

```suggestion
export type SidebarSearchModalProps = Pick<
  SearchModalProps,
  'children' | 'resultItems'
> & {
  icon?: IconComponent;
};
```

File: plugins/search/src/components/SidebarSearchModal/SidebarSearchModal.tsx
Comment (camilaibs) [inline plugins/search/src/components/SidebarSearchModal/SidebarSearchModal.tsx:24]:
```suggestion
  useSearchModal,
  SearchModalProps,
```

File: plugins/search/src/components/SidebarSearchModal/SidebarSearchModal.tsx
Comment (camilaibs) [inline plugins/search/src/components/SidebarSearchModal/SidebarSearchModal.tsx:53]:
```suggestion
        resultItems={props.resultItems}
```

File: plugins/search/src/index.ts
Comment (camilaibs) [inline plugins/search/src/index.ts:46]:
```suggestion

```

File: plugins/search/src/plugin.ts
Comment (camilaibs) [inline plugins/search/src/plugin.ts:39]: I recommend defining this in the props of the component where it is used, in this case, the `SearchModalProps` type.

As mentioned in the other comments people completely customizing the search modal will define search results manually and the `SidebarSearchModalProps` can pick the `resultItems` type from the `SearchModalProps`.

```suggestion
import { SidebarSearchModalProps } from './components/SidebarSearchModal';

```

**Review by camilaibs — CHANGES_REQUESTED** [review]
Thank you so much for contributing with us, It makes it easy to have custom items in the default search modal 🎉 

I left comments about 2 minor suggestions: the name of the new prop and where to define it. Please take a look when you have a moment 🙏🏻 

The documentation looks amazing 👏🏻

File: docs/features/search/how-to-guides.md
Comment (joshjung) [reply docs/features/search/how-to-guides.md:301]: So my reasoning for calling it `searchResultChildren` is because it is the `children` of the `SearchResult` component. I wanted that to be explicit, so that the next devs know exactly where the children will go. I can certainly switch it but personally like really explicit property names.

File: plugins/search/src/components/SearchModal/SearchModal.tsx
Comment (joshjung) [reply plugins/search/src/components/SearchModal/SearchModal.tsx:55]: I think I see what you're saying, will take a look

File: plugins/search/src/components/SearchModal/SearchModal.tsx
Comment (joshjung) [reply plugins/search/src/components/SearchModal/SearchModal.tsx:105]: I appreciate what you are saying, but also feel a bit compelled to push back on this a bit again, I think that being clear that we are passing the children to the SearchResult component is important here. Thoughts?

File: plugins/search/src/components/SidebarSearchModal/SidebarSearchModal.tsx
Comment (joshjung) [reply plugins/search/src/components/SidebarSearchModal/SidebarSearchModal.tsx:36]: But what if you want to provide both?

File: plugins/search/src/components/SearchModal/SearchModal.tsx
Comment (joshjung) [reply plugins/search/src/components/SearchModal/SearchModal.tsx:45]: Oh I get it now

File: docs/features/search/how-to-guides.md
Comment (camilaibs) [reply docs/features/search/how-to-guides.md:301]: I see your point, let me share the reasoning behind my name suggestion, but feel free to decide 🙂:
I think we don't need to necessarily couple the property name with the implementation of the search modal, I mean, to me people using this component don't need to know how the property will be used internally by the search modal code. If the ‘SearchResult’ gets replaced in the search modal implementation in the future or stop receiving list items via children the name ´searchResultChildren’ will no longer make sense, but result items continue being a meanful name because it will continue be receiving a list of items independently if it is forwarded as a children or not.

File: plugins/search/src/components/SearchModal/SearchModal.tsx
Comment (camilaibs) [reply plugins/search/src/components/SearchModal/SearchModal.tsx:105]: Replied in the first comment, let me know what do you think 🙂

File: plugins/search/src/components/SidebarSearchModal/SidebarSearchModal.tsx
Comment (camilaibs) [reply plugins/search/src/components/SidebarSearchModal/SidebarSearchModal.tsx:36]: In this case we are picking both children and resulItems, not sure if I understand the question

File: docs/features/search/how-to-guides.md
Comment (joshjung) [reply docs/features/search/how-to-guides.md:301]: Okay I'll switch it, I get what you are saying now. I think both have benefits but also both make sense. The important thing is people can choose.

Comment (joshjung) [reply docs/features/search/how-to-guides.md:301]: I have a compromise suggestion. Rather than `resultItems` (since I would read this as the actual results themselves), how about `resultItemComponents`?

File: plugins/search/src/components/SidebarSearchModal/SidebarSearchModal.tsx
Comment (joshjung) [reply plugins/search/src/components/SidebarSearchModal/SidebarSearchModal.tsx:36]: Sorry, I misunderstood what you meant. I was still under the impression that you could provide children and resultItemComponents

File: docs/features/search/how-to-guides.md
Comment (camilaibs) [reply docs/features/search/how-to-guides.md]: That's a good idea, I like ’resultItemComponents’ 🙌

File: plugins/search/src/components/SearchModal/SearchModal.tsx
Comment (camilaibs) [inline plugins/search/src/components/SearchModal/SearchModal.tsx:233]: I believe it is not necessary to pass an empty array here because resultItemComponents can be undefined.

**Review by camilaibs — APPROVED** [review]
Looks great to me, thanks again for opening the pull request 🎉 
@backstage/search-maintainers could you please take a look?

File: plugins/search/src/components/SearchType/SearchType.Tabs.tsx
Comment (emmaindal) [inline plugins/search/src/components/SearchType/SearchType.Tabs.tsx:23]: Can we skip using `any` here and use `Theme` instead from `@material-ui/core/styles`

**Review by emmaindal — COMMENTED** [review]
Thank you for this contribution, and thanks @camilaibs for great reviewing! 

Looks good to me! Just one small comment!

File: plugins/search/src/components/SearchType/SearchType.Tabs.tsx
Comment (joshjung) [reply plugins/search/src/components/SearchType/SearchType.Tabs.tsx:23]: Oh yeah for sure! Sorry I missed this.

**Review by emmaindal — APPROVED** [review]
Thanks a lot 🎉

**Comment by joshjung** [comment]
@camilaibs Base branch does not allow updates, merging is blocked?

**Comment by camilaibs** [comment]
> @camilaibs Base branch does not allow updates, merging is blocked?

Yes, the `master` branch is blocked until we get the todays release out.

**Pull request merged**
